e2e/service_test.go

```diff
@@ -0,0 +1,197 @@
+//go:build e2e
+
+package e2e
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/buildwithgrove/path/gateway"
+	"github.com/buildwithgrove/path/protocol"
+	"github.com/buildwithgrove/path/qos/jsonrpc"
+	"github.com/buildwithgrove/path/request"
+	vegeta "github.com/tsenart/vegeta/lib"
+)
+
+type serviceType string
+
+const (
+	serviceTypeEVM      serviceType = "evm"
+	serviceTypeCometBFT serviceType = "cometbft"
+	serviceTypeSolana   serviceType = "solana"
+)
+
+// -----------------------------------------------------------------------------
+// TestServices Struct - Configures the services to test against.
+//
+// Unmarshalled from the YAML files:
+//   - `config/services_morse.yaml`
+//   - `config/services_shannon.yaml`
+// -----------------------------------------------------------------------------
+
+// DEV_NOTE: All structs and `yaml:` tagged fields must be public to allow for unmarshalling using `gopkg.in/yaml`
+type (
+	TestServices struct {
+		Services []TestService `yaml:"services"` // List of test services to run the tests against
+	}
+
+	TestService struct {
+		Name          string             `yaml:"name"`               // Name of the service
+		ServiceID     protocol.ServiceID `yaml:"service_id"`         // Service ID to test (identifies the specific blockchain service)
+		ServiceType   serviceType        `yaml:"service_type"`       // Type of service to test (evm, cometbft, solana)
+		Archival      bool               `yaml:"archival,omitempty"` // Whether this is an archival test (historical data access)
+		ServiceParams ServiceParams      `yaml:"service_params"`     // Service-specific parameters for test requests
+	}
+
+	// ServiceParams holds service-specific test data for all methods.
+	// TODO_IMPROVE(@commoddity): Look into getting contract address and contract start block
+	// from `config/service_qos_config.go` to have only one source of truth for service params
+	ServiceParams struct {
+		ContractAddress    string `yaml:"contract_address,omitempty"`     // EVM contract address (should match service_qos_config.go)
+		ContractStartBlock uint64 `yaml:"contract_start_block,omitempty"` // Minimum block number to use for archival tests
+		TransactionHash    string `yaml:"transaction_hash,omitempty"`     // Transaction hash for receipt/transaction queries
+		CallData           string `yaml:"call_data,omitempty"`            // Call data for eth_call
+		blockNumber        string // Not marshaled from YAML; set in test case. Can be "latest" or an archival block number
+	}
+)
+
+func (ts *TestService) getTestMethods() []string {
+	switch ts.ServiceType {
```

Comment:
Not a blocker: Interfaces would significantly simplify this: the QoS implementation for the service should be handling this, even if it is as simple as a public method like `JSONRPCMethodsToTest`

Reply:
Also agreed here. Based on our discussion today we will handle this change by moving as much test logic as possible to the `qos` packages.